crates/red_knot_python_semantic/src/types.rs

```diff
@@ -867,12 +885,76 @@ impl<'db> IterationOutcome<'db> {
             Self::Iterable { element_ty } => element_ty,
             Self::NotIterable { not_iterable_ty } => {
                 inference_builder.not_iterable_diagnostic(iterable_node, not_iterable_ty);
-                Type::Unknown
+                Type::Unknown(UnknownTypeKind::TypeError)
             }
         }
     }
 }
 
+// When it really matters to distinguish between different kinds of `Unknown` types
+// (see `PartialEq` implementation), we use the representation as `u8`
+#[derive(Debug, Clone, Copy, Eq)]
+#[repr(u8)]
+pub enum UnknownTypeKind {
+    /// Temporary variant that indicates that we *should*
+    /// be able to infer a type here in due course, but currently can't
+    RedKnotLimitation,
+
+    /// Invalid syntax in the node means we can't infer a type here
+    InvalidSyntax,
+
+    /// An `Unknown` type stemming from some kind of type error.
+    ///
+    /// Examples:
+    /// - A function was called with argument(s) of incorrect type(s), leading
+    ///   to a diagnostic being emitted and the expression being evaluated as `Unknown`.
+    /// - An expression was deemed to take place between two invalid types, e.g. `1 + "foo"`.
+    ///   (This is really the same as the first example, since expressions ultimately desguar
+    ///   to function calls, e.g. `1 + "foo"` desugars approximately to `type(1).__add__("foo")`.)
+    /// - An attribute or method was looked up on an instance of a type that doesn't have that
+    ///   attribute or method.
+    TypeError,
+
+    /// The symbol was unannotated and the true type can't be reasonably inferred
+    UnannotatedSymbol,
+
+    /// The type of symbols imported by imports that could not be resolved
+    UnresolvedImport,
+
+    /// A "second-order" Unknown type, that results from an interaction with
+    /// a "first-order" Unknown type. For example, if the type of `x` is `Unknown`,
+    /// the type of `x + 1` will also be `Unknown`
+    SecondOrder,
+
+    /// Special kind for intersections that is equal to all other kinds. This allows to consider a
+    /// single `Type::Unknown(_)` in intersections. Should not be used in other contexts.
+    AllKinds,
```

Comment:
I'm also not convinced we should have this; I think I'd prefer to just not simplify differing unknowns in intersections. I don't expect this case to arise often, because most intersections will come from narrowing constraints, and in most cases we'll just not return a narrowing constraint at all rather than return one with an unknown type.

Reply:
To be clear, does that mean that having potentially different unknown kinds inside an intersection is not a concern?

Reply:
Yeah, I don't think that's a problem.